python/openmldb/dbapi/dbapi.py

```diff
@@ -299,9 +275,93 @@ def execute(self, operation, parameters=()):
+            sql_router_sdk.kTypeDouble: builder.AppendDouble,
+            sql_router_sdk.kTypeString: builder.AppendString,
+            # TODO: align python and java date process, 1900 problem
+            sql_router_sdk.kTypeDate: lambda x: len(x.split("-")) == 3 and builder.AppendDate(
+                int(x.split("-")[0]), int(x.split("-")[1]), int(x.split("-")[2])),
+            sql_router_sdk.kTypeTimestamp: builder.AppendTimestamp
+        }
+        return appendMap
+
+    def __insert_rows(self, rows: List[Union[tuple, dict]], hold_idxs, schema, rows_builder, command):
+        for row in rows:
+            tmp_builder = rows_builder.NewRow()
+            appendMap = self.__get_append_map(tmp_builder, row, hold_idxs, schema)
+            self.__add_row_to_builder(row, hold_idxs, schema, tmp_builder, appendMap)
+        ok, error = self.connection._sdk.executeInsert(self.db, command, rows_builder)
+        if not ok:
+            raise DatabaseError(error)
+
     @connected
-    def executemany(self, operation, parameters=()):
-        raise NotSupportedError("Unsupported in OpenMLDB")
+    def executemany(self, operation, parameters: Union[List[tuple], List[dict]], batch_number=200):
+        parameters_length = len(parameters)
+        command = operation.strip(' \t\n\r') if operation else None
+        if command is None:
+            raise Exception("None operation")
+        if command.count("?") == 0:
+            logging.warning(
+                "Only {} is valid, params: {} are invalid, maybe not exists mark '?' in sql".format(operation,
+                                                                                                    parameters))
+            return self.execute(operation, parameters)
+        if isinstance(parameters, list) and parameters_length == 0:
+            return self.execute(operation, parameters)
+
+        if insertRE.match(command):
+            question_mark_count = command.count("?")
+            if question_mark_count > 0:
+                # Because the object obtained by getInsertBatchBuilder has no GetSchema method,
+                # use the object obtained by getInsertBatchBuilder
+                ok, builder = self.connection._sdk.getInsertBuilder(self.db, command)
+                if not ok:
+                    raise DatabaseError("get insert builder fail")
+                schema = builder.GetSchema()
+                hold_idxs = builder.GetHoleIdx()
+                for i in range(0, parameters_length, batch_number):
+                    rows = parameters[i: i + batch_number]
+                    ok, batch_builder = self.connection._sdk.getInsertBatchBuilder(self.db, command)
+                    if not ok:
+                        raise DatabaseError("get insert builder fail")
+                    self.__insert_rows(rows, hold_idxs, schema, batch_builder, command)
+            else:
+                ok, error = self.connection._sdk.executeInsert(self.db, command)
```

Comment:
we have `execute_sql(db, sql)` API that support all kinds of SQL without input params.

Reply:
Do I need to make changes?

Reply:
I suppose so.

Reply:
ok